crates/miden-objects/src/transaction/script_builder.rs

```diff
@@ -0,0 +1,314 @@
+/// 3. Compile your note script with `compile_note_script()`
+/// 4. Compile your transaction script with `compile_tx_script()`
+///
+/// # Note
+/// For scripts that need access to transaction kernel procedures,
+/// use `ScriptBuilder::with_assembler()` and provide an assembler created with
+/// `TransactionKernel::assembler()` from the miden-lib crate.
+pub struct ScriptBuilder {
+    assembler: Assembler,
+    libraries: Vec<Library>,
+}
+
+impl ScriptBuilder {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new ScriptBuilder with the specified debug mode.
+    ///
+    /// This creates a basic assembler. For transaction scripts that need access to
+    /// transaction kernel procedures, use `ScriptBuilder::with_assembler()` and provide
+    /// an assembler created with `TransactionKernel::assembler()`.
+    ///
+    /// # Arguments
+    /// * `in_debug_mode` - Whether to enable debug mode in the assembler
+    pub fn new(in_debug_mode: bool) -> Self {
+        let assembler = Assembler::default().with_debug_mode(in_debug_mode);
+        Self { assembler, libraries: Vec::new() }
+    }
+
+    /// Creates a new ScriptBuilder with a provided assembler.
+    ///
+    /// This is the recommended constructor when you need access to transaction kernel
+    /// procedures. Pass `TransactionKernel::assembler()` as the assembler parameter.
+    ///
+    /// # Arguments
+    /// * `assembler` - A pre-configured assembler (e.g., from TransactionKernel::assembler())
+    /// * `in_debug_mode` - Whether to enable debug mode
+    pub fn with_assembler(assembler: Assembler, in_debug_mode: bool) -> Self {
+        let assembler = assembler.with_debug_mode(in_debug_mode);
+        Self { assembler, libraries: Vec::new() }
+    }
+
+    // LIBRARY MANAGEMENT
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds multiple libraries to the script builder.
+    ///
+    /// # Arguments
+    /// * `libraries` - Vector of compiled libraries to add to the builder
+    pub fn add_library(&mut self, libraries: Vec<Library>) -> Result<(), ScriptBuilderError> {
+        for library in libraries {
+            self.libraries.push(library);
+        }
+        Ok(())
+    }
+
+    // SCRIPT COMPILATION
+    // --------------------------------------------------------------------------------------------
+
+    /// Compiles a transaction script with the provided program code.
+    ///
+    /// The compiled script will have access to all libraries that have been added to this builder.
+    ///
+    /// # Arguments
+    /// * `program` - The transaction script source code
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - Any of the libraries cannot be loaded into the assembler
+    /// - The transaction script compilation fails
+    pub fn compile_tx_script(
+        &self,
+        tx_script: &str,
```

Comment:
Is there a reason not to use `impl AsRef<str>` instead of `&str` here? I think it would cover all the use cases as `&str` does but would be a bit more flexible.